File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Tuple{
    int num;
    TreeNode node;
    public Tuple(TreeNode node, int num){
        this.node = node;
        this.num = num;
    }
}

class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        int[] arr = {0};
        Queue<Tuple> queue = new LinkedList<>();
        TreeMap<Integer,PriorityQueue<Integer>> map = new TreeMap<>();
        queue.offer(new Tuple(root,0));
        int ans = 0;
        while(!queue.isEmpty()){
            Tuple tuple = queue.peek();
            TreeNode node = tuple.node;
            int mmin = tuple.num;
            int size = queue.size();
            int first = 0, last = 0;
            for(int i=0;i<size;i++){
                int cur_id = queue.peek().num-mmin;
                node = queue.poll().node;
                if(i==0) first = cur_id;
                if(i==size-1) last = cur_id;
                if(node.left!=null) queue.offer(new Tuple(node.left,cur_id*2+1));
                if(node.right!=null) queue.offer(new Tuple(node.right,cur_id*2+2));
                }
            ans = Math.max(ans, last-first+1);
            }
        return ans;
    }
}

File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        ansdict = defaultdict(list)
        todo = deque([(root, 0, 0)])
        while todo:
            node, level, view = todo.popleft()
            ansdict[level].append(view)
            view = view-1
            if node.left:
                todo.append((node.left,level+1,2*view + 1))
            if node.right:
                todo.append((node.right,level+1, 2*view + 2))
        ans = 0
        for i in sorted(ansdict):
            ans  = max(ans, max(ansdict[i])-min(ansdict[i])+1)
        return ans